2020-01-Winter/0_arrays_sorting/README.md

```diff
@@ -94,21 +105,122 @@ Go to [Solution](#s3)   [Top](#top)
+        Insert nums2 into current posistion
+        Increment current nums1 character counter
+        Increment nums2 tracker
+    Otherwise, go through the array until you reach end of nums1 character counter
+        All nums2 larger than nums1 at this point
+        Insert all nums2 at end
+Return merged array
+```
+This solution is space optimal, with O(1) space. However, this solution is not time optimal. For each insertion operation you make, you will have to push back N number of objects, N number of times. This implementation would have a runtime of O(N^2)
+
+Example code of this implementation:
+```
+std::vector<int> &merge(std::vector<int> &nums1, int m, std::vector<int> &nums2, int n)
+{
+  int tracker = 0;
+  
+  for(int i = 0; i < nums1.size(); i++)
+  {
+    if(nums2[tracker] < nums1[i] || nums2[tracker] == nums1[i])
+    {
+      for(int k = m-1; k >= i; k--)
+      {
+        nums1[k+1] = nums1[k];
+      }
+    
+      nums1[i] = nums2[tracker];
+      m++;
+      tracker++;
+    }
+    else if (i > m-1)
+    {
+      nums1[i] = nums2[tracker];
+      tracker++;
+    }
+  }
+
+  return nums1;
+}
+```
+
+
+#### Optimal solution
+
+A better solution can hit both the space requirements, while being time optimal. It requires starting from the back of the array, rather than stepping forward from the front, and is expressed in the following psuedocode:
+```
+Start from the back of the nums1 array
+
+While the number of characters nums1 has stored is greater than 0
+    Check if the last character of nums1 is smaller than nums2 last character
+        If true, place nums2 last element in the back of the array
+        Decrement nums2 character counter
+        Set back of the array to be before the newly added character
+    Otherwise, check if nums1 last character is bigger than nums2 last character
+        If so, move nums2 last character to back of array
+        Set nums1 previous position to 0
+        Decrement nums1 character counter
+        Set back of the array to be before the newly added character
+    Otherwise, nums1 and nums 2 characters are the same
+        Set back of array equal to nums2 last charact
+        Decrement both nums1 and nums2 character counter
+When all done, return the newly merge nums1
+
+``` 
+
+This implementation has a runtime of O(N), and space complexity of O(1). The full implementation in C++ is as follows:
+
+```
+std::vector<int>& merge(std::vector<int>& nums1, int m, std::vector<int>& nums2, int n) 
+{
+  int backOfArray = nums1.size()-1;
+
+  while(m-1 > 0)
+  {
+    if(nums1[m-1] < nums2[n-1])
+    {
+      nums1[backOfArray] = nums2[n-1];
+      std::cout << n;
+      n--;
+      backOfArray--;
+    }
+    else if (nums1[m-1] > nums2[n-1])
+    {
+      nums1[backOfArray] = nums1[m-1];
+      nums1[m-1] = 0;
+
+      m--;
+      backOfArray--;
+    }
+    else
+    {
+      nums1[backOfArray] = nums2[n-1];
+      n--;
+      m--;
```

Comment:
Remove this line to fix the logical error (see comment in pseudocode section above, line 178-180)
```suggestion
```